File: bbtest/steps/ledger_steps.rb
require_relative 'placeholders'

require 'json'
require 'bigdecimal'

step "following transaction is created from tenant :tenant" do |tenant, body = nil|
  send "following transaction is created :count times from tenant :tenant", 1, tenant, body
end

step "following transaction is created :count times from tenant :tenant" do |count, tenant, body = nil|
  uri = "https://127.0.0.1:4401/transaction/#{tenant}"

  send "I request curl :http_method :url", "POST", uri, body

  [*1..count].each { |_|
    eventually() {
      send "curl responds with :http_status", [200, 201]
    }
  }
end

step ":id :id :side side is forwarded to :account from tenant :tenant" do |transaction, transfer, side, account, tenant|
  (tenant, account) = account.split('/')

  payload = {
    side: side,
    target: {
      tenant: tenant,
      name: account
    }
  }.to_json

  uri = "https://127.0.0.1:4401/transaction/#{tenant}/#{transaction}/#{transfer}"

  send "I request curl :http_method :url", "PATCH", uri, payload
  send "curl responds with :http_status", [200, 201]
end

step ":amount :currency is transferred from :account to :account" do |amount, currency, from, to|
  (fromTenant, fromAccount) = from.split('/')
  (toTenant, toAccount) = to.split('/')

  payload = {
    transfers: [{
      credit: {
        name: toAccount,
        tenant: toTenant,
      },
      debit: {
        name: fromAccount,
        tenant: fromTenant,
      },
      amount: amount,
      currency: currency
    }]
  }.to_json

  send "following transaction is created from tenant :tenant", fromTenant, payload
end

File: bbtest/steps/ledger_steps.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from behave import *
import json
import time
from helpers.http import Request, StubResponse


@given('{amount} {currency} is transferred from {tenantFrom}/{accountFrom} to {tenantTo}/{accountTo}')
@when('{amount} {currency} is transferred from {tenantFrom}/{accountFrom} to {tenantTo}/{accountTo}')
def create_transaction(context, amount, currency, tenantFrom, accountFrom, tenantTo, accountTo):
  uri = "https://127.0.0.1:4401/transaction/{}".format(tenantFrom)
  payload = {
    'transfers': [{
      'credit': {
        'name': accountTo,
        'tenant': tenantTo,
      },
      'debit': {
        'name': accountFrom,
        'tenant': tenantFrom,
      },
      'amount': amount,
      'currency': currency
    }]
  }

  request = Request(method='POST', url=uri)
  request.add_header('Content-Type', 'application/json')
  request.add_header('Accept', 'application/json')
  request.data = json.dumps(payload)

  response = StubResponse(202, '')

  while response.status == 202:
    response = request.do()

  assert response.status in [200, 201], 'expected status 200 or 201 actual {}'.format(response.status)


@given('following transaction is created {times} times from tenant {tenant}')
@when('following transaction is created {times} times from tenant {tenant}')
def create_transaction_literal_times(context, times, tenant):
  uri = "https://127.0.0.1:4401/transaction/{}".format(tenant)

  request = Request(method='POST', url=uri)
  request.data = context.text
  request.add_header('Content-Type', 'application/json')
  request.add_header('Accept', 'application/json')

  for _ in range(int(times)):
    
    response = StubResponse(202, '')

    while response.status == 202:
      response = request.do()

    assert response.status in [200, 201], 'expected status 200 or 201 actual {}'.format(response.status)


@given('following transaction is created from tenant {tenant}')
@when('following transaction is created from tenant {tenant}')
def create_transaction_literal(context, tenant):
  create_transaction_literal_times(context, 1, tenant)
